ragna/core/_rag.py

```diff
@@ -237,42 +278,64 @@ def _parse_documents(self, documents: Iterable[Any]) -> list[Document]:
 
     def _unpack_chat_params(
         self, params: dict[str, Any]
-    ) -> dict[tuple[Type[Component], str], dict[str, Any]]:
-        source_storage_models = self.source_storage._protocol_models()
-        assistant_models = self.assistant._protocol_models()
+    ) -> dict[Callable, dict[str, Any]]:
+        component_models = {
+            getattr(component, name): model
+            for component in [self.source_storage, self.assistant]
+            for (_, name), model in component._protocol_models().items()
+        }
 
         ChatModel = merge_models(
             str(self.params["chat_id"]),
             SpecialChatParams,
-            *source_storage_models.values(),
-            *assistant_models.values(),
+            *component_models.values(),
             config=pydantic.ConfigDict(extra="forbid"),
         )
 
         chat_params = ChatModel.model_validate(params, strict=True).model_dump(
             exclude_none=True
         )
         return {
-            component_and_action: model(**chat_params).model_dump()
-            for component_and_action, model in itertools.chain(
-                source_storage_models.items(), assistant_models.items()
-            )
+            fn: model(**chat_params).model_dump()
+            for fn, model in component_models.items()
         }
 
-    async def _enqueue(
-        self, component: Type[Component], action: str, *args: Any
-    ) -> Any:
-        try:
-            return await self._rag._queue.enqueue(
-                component, action, args, self._unpacked_params[(component, action)]
+    def _run_async(self, fn: Awaitable[T]) -> T:
+        # This is the best way I could come up with to be able to run an async function
+        # synchronously when there is already an event-loop running. This happens for
+        # example inside a jupyter notebook. Without this, it would be impossible to use
+        # synchronous methods in such a scenario.
+        # FIXME: That being said, it feels kinda hacky and is also causing issues in the
+        #  test suite. If anyone knows a better way to achieve this, please send a PR :)
+        nest_asyncio.apply()
+        return asyncio.run(fn)  # type: ignore[arg-type]
```

Comment:
This is really a bummer :disappointed: We have this in the first place, since the actual functionality is implemented in our `async` endpoints, e.g. `aanswer`. To avoid re-implementing the same body for the sync counterpart, I just run the `async` endpoint here. However, we need to resort to monkey-patching (it is done with care in `nest_asyncio`, but ultimately that is what it is) to do this when another event loop is already running. This happens for example in a jupyter notebook and thus is critical feature.